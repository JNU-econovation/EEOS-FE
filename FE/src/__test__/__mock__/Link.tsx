export const Link = jest.fn(({ children, href }) => (
  <a role="link" href={href} onClick={(e) => e.preventDefault()}>
    {children}
  </a>
));
