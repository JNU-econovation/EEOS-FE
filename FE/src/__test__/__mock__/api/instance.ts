export const https = jest.fn();
