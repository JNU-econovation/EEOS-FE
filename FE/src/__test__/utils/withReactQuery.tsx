import { QueryClient, QueryClientProvider } from "@tanstack/react-query";

const queryClient = new QueryClient();

const withReactQuery = (Component: React.ReactNode) => {
  return () => (
    <QueryClientProvider client={queryClient}>{Component}</QueryClientProvider>
  );
};

export default withReactQuery;
