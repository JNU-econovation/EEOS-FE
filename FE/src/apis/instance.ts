import axios from "axios";
import { toast } from "react-toastify";
import { postTokenReissue } from "./auth";
import ERROR_CODE from "@/constants/ERROR_CODE";
import ERROR_MESSAGE from "@/constants/ERROR_MESSAGE";
import {
  deleteTokenInfo,
  getAccessToken,
  getTokenExpiration,
  setAccessToken,
  setTokenExpiration,
} from "@/utils/authWithStorage";
import ROUTES from "@/constants/ROUTES";

const https = axios.create({
  baseURL: process.env.NEXT_PUBLIC_API_URL + "/api",
  headers: {
    "Content-Type": "application/json",
  },
  withCredentials: true,
});

const authInstance = axios.create({
  baseURL: process.env.NEXT_PUBLIC_API_URL + "/api",
  headers: {
    "Content-Type": "application/json",
  },
  withCredentials: true,
});

https.interceptors.request.use(
  async (config) => {
    if (typeof window === "undefined") return config;
    const accessToken = getAccessToken();
    const tokenExpiration = getTokenExpiration();

    if (!accessToken || !tokenExpiration) return config;

    const currentTime = new Date().getTime();
    const timeToExpiration = Number(tokenExpiration) - currentTime;
    const TOKEN_REISSUE_THRESHOLD = Number(
      process.env.NEXT_PUBLIC_TOKEN_REISSUE_THRESHOLD,
    );

    if (timeToExpiration < TOKEN_REISSUE_THRESHOLD) {
      try {
        const { accessToken, accessExpiredTime } = await postTokenReissue();
        setAccessToken(accessToken);
        setTokenExpiration(accessExpiredTime);
      } catch (e) {
        deleteTokenInfo();
        toast.error(ERROR_MESSAGE[ERROR_CODE.AUTH.INVALID_TOKEN].message, {
          toastId: ERROR_CODE.AUTH.INVALID_TOKEN,
        });

        setTimeout(() => {
          window.location.href = ROUTES.LOGIN;
        }, 3000);
      }
    }

    config.headers["Authorization"] = `Bearer ${accessToken}`;

    return config;
  },
  (error) => Promise.reject(error),
);

https.interceptors.response.use(
  async (config) => {
    return config;
  },
  async (error) => {
    const { config: originalRequest, response } = error;
    const errorCode = response?.data?.code;
    const errorMessage =
      ERROR_MESSAGE[errorCode]?.message || ERROR_MESSAGE.UNKNOWN.message;

    if (errorCode === ERROR_CODE.AUTH.INVALID_TOKEN) {
      deleteTokenInfo();
      toast.error(errorMessage, {
        toastId: errorCode,
      });
      setTimeout(() => {
        window.location.href = ROUTES.LOGIN;
      }, 3000);
    }

    if (errorCode === ERROR_CODE.AUTH.EXPIRED_ACCESS_TOKEN) {
      const { accessToken, accessExpiredTime } = await postTokenReissue();

      setAccessToken(accessToken);
      setTokenExpiration(accessExpiredTime);
      originalRequest.headers["Authorization"] = `Bearer ${accessToken}`;

      return await axios(originalRequest);
    }

    if (errorCode === ERROR_CODE.AUTH.INVALID_NAME) {
      toast.error(errorMessage, {
        toastId: errorCode,
      });
      return Promise.reject(error);
    }

    if (Object.values(ERROR_CODE.AUTH).includes(errorCode)) {
      toast.error(errorMessage, {
        toastId: errorCode,
      });
      deleteTokenInfo();
      setTimeout(() => {
        if (errorCode === ERROR_CODE.AUTH.INCORRECT_LOGIN_INFO) return;
        window.location.href = "/login";
      }, 3000);
    }
    error.message = errorMessage;
    return Promise.reject(error);
  },
);

export { https, authInstance };
