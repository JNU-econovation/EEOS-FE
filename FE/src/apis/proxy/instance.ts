import axios from "axios";

const proxy = axios.create({
  baseURL: "/api",
});

export default proxy;
