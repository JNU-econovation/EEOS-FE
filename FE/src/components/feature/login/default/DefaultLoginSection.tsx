import GuestLoginBtn from "./GuestLoginBtn";
import SlackLoginBtn from "./SlackLoginBtn";
import Title from "@/components/common/Title/Title";
import { ArrowRight } from "@/components/icons";

interface DefaultLoginSectionProps {
  changeLoginType: () => void;
}
const DefaultLoginSection = ({ changeLoginType }: DefaultLoginSectionProps) => {
  return (
    <div className="flex flex-col items-center justify-center gap-24">
      <Title text={"로그인"} />
      <div className="flex flex-col gap-6">
        <div className="flex flex-col items-center gap-6">
          <p className="font-light">에코노베이션 슬랙으로 로그인</p>
          <SlackLoginBtn />
        </div>
        <div className="flex flex-col items-center gap-6">
          <p className="font-light">게스트모드로 EEOS 둘러보기</p>
          <GuestLoginBtn />
        </div>
        <p
          onClick={() => changeLoginType()}
          className="mx-auto flex select-none"
        >
          관리자 로그인
          <ArrowRight />
        </p>
      </div>
    </div>
  );
};

export default DefaultLoginSection;
