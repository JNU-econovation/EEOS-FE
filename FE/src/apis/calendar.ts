import { SimpleCalendarDto } from "@/apis/dtos/calendar.dto";
import { https } from "@/apis/instance";
import API from "@/constants/API";
import { DateFilter, NewCalendar } from "@/types/calendar";

export const getCalendarEventsOnWeek = async (
  dateFilter: DateFilter,
): Promise<SimpleCalendarDto[]> => {
  const { data } = await https({
    method: "GET",
    url: API.CALENDAR.WEEKLY(dateFilter),
  });

  return data.data.calendars.map(
    (item: SimpleCalendarDto) => new SimpleCalendarDto({ ...item }),
  );
};

export function postCalender(newCalendar: NewCalendar) {
  return https({
    url: API.CALENDAR.CREATE,
    method: "POST",
    data: newCalendar,
  });
}

export function deleteCalender(calendarId: number) {
  return https({
    url: API.CALENDAR.DELETE(calendarId),
    method: "DELETE",
  });
}
