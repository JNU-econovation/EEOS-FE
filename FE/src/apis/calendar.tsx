import { SimpleCalendarDto } from "@/apis/dtos/calendar.dto";
import { https } from "@/apis/instance";
import API from "@/constants/API";

export const getCalendarEventsOnWeek = async (): Promise<
  SimpleCalendarDto[]
> => {
  const { data } = await https({
    method: "GET",
    url: API.CALENDAR.WEEKLY({
      year: new Date().getFullYear(),
      month: new Date().getMonth() + 1,
      date: new Date().getDate(),
      duration: 14,
    }),
  });

  return data.data.calendars.map(
    (item: SimpleCalendarDto) => new SimpleCalendarDto({ ...item }),
  );
};
