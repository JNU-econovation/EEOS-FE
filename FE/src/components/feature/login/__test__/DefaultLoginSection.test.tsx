/**
 * @jest-environment jsdom
 */

import { render, screen } from "@testing-library/react";
import renderer from "react-test-renderer";
import DefaultLoginSection from "../default/DefaultLoginSection";
import { Link } from "@/__test__/__mock__/Link";
import getResponse from "@/__test__/__stub__/response";
import withReactQuery from "@/__test__/utils/withReactQuery";
import { https } from "@/apis/instance";

//axios mocking
jest.mock("@/apis/instance");
const mockHttps = https as jest.MockedFunction<typeof https>;

const mockReturnData = getResponse({
  url: "/auth/login/slack",
  method: "POST",
});

mockHttps.mockResolvedValue(mockReturnData);

// next.js mocking
const MockNextRouterComponent = Link;

jest.mock("next/navigation", () => ({
  useSearchParams: jest.fn(() => ({
    get: jest.fn(() => "code"),
  })),
  useRouter: jest.fn(() => ({
    back: jest.fn(),
    replace: jest.fn(),
  })),
}));

jest.mock("next/link", () => {
  return ({ children, href }) => {
    return (
      <MockNextRouterComponent href={href}>{children}</MockNextRouterComponent>
    );
  };
});

// module mocking
const changeLoginType = jest.fn();

describe("DefaultLoginSection", () => {
  it("올바르게 렌더링 된다", async () => {
    const WrappedComponent = withReactQuery(
      <DefaultLoginSection changeLoginType={changeLoginType} />,
    );
    const tree = renderer.create(<WrappedComponent />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("게스트 모드로 이동할 수 있다", async () => {
    const WrappedComponent = withReactQuery(
      <DefaultLoginSection changeLoginType={changeLoginType} />,
    );
    render(<WrappedComponent />);

    const guestLoginButton = screen.getByRole("link", {
      name: "Visit EEOS",
    });

    guestLoginButton.click();

    expect(MockNextRouterComponent).toHaveBeenCalled();
  });

  it("관리자 로그인 버튼 클릭시 로그인 타입을 변경한다", () => {
    const WrappedComponent = withReactQuery(
      <DefaultLoginSection changeLoginType={changeLoginType} />,
    );
    render(<WrappedComponent />);

    const adminLoginSpan = screen.getByText("관리자 로그인");

    adminLoginSpan.click();

    expect(changeLoginType).toHaveBeenCalledTimes(1);
  });
});
