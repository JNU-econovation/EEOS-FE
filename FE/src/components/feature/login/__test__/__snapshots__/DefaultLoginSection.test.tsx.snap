// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DefaultLoginSection 올바르게 렌더링 된다 1`] = `
<div
  className="flex flex-col items-center justify-center gap-24"
>
  <h1
    className="text-2xl font-bold sm:text-3xl text-black"
  >
    로그인
  </h1>
  <div
    className="flex flex-col gap-6"
  >
    <div
      className="flex flex-col items-center gap-6"
    >
      <p
        className="font-light"
      >
        에코노베이션 슬랙으로 로그인
      </p>
      <a
        href="https://slack.com/oauth/v2/authorize?client_id=5775976606341.6306752698400&team_id=undefined&scope=&user_scope=users.profile:read&redirect_uri=https://localhost:3000/login/logging-in"
        onClick={[Function]}
        role="link"
      >
        <img
          alt="슬랙 로그인 버튼"
          data-nimg="1"
          decoding="async"
          height={24}
          loading="lazy"
          onError={[Function]}
          onLoad={[Function]}
          src="/icons/slack.svg"
          style={
            {
              "color": "transparent",
            }
          }
          width={24}
        />
        <p
          className="text-center font-semibold"
        >
          슬랙으로 로그인
        </p>
      </a>
    </div>
    <div
      className="flex flex-col items-center gap-6"
    >
      <p
        className="font-light"
      >
        게스트모드로 EEOS 둘러보기
      </p>
      <a
        href="/guest/main"
        onClick={[Function]}
        role="link"
      >
        <img
          alt=""
          data-nimg="1"
          decoding="async"
          height={24}
          loading="lazy"
          onError={[Function]}
          onLoad={[Function]}
          src="/icons/blackCompany.svg"
          style={
            {
              "color": "transparent",
            }
          }
          width={24}
        />
        <p
          className="text-center font-semibold"
        >
          Visit EEOS
        </p>
      </a>
    </div>
    <p
      className="mx-auto flex select-none"
      onClick={[Function]}
    >
      관리자 로그인
      <img
        alt="arrow right to go admin login page"
        data-nimg="1"
        decoding="async"
        height={20}
        loading="lazy"
        onError={[Function]}
        onLoad={[Function]}
        src="/icons/arrowRight.svg"
        style={
          {
            "color": "transparent",
          }
        }
        width={20}
      />
    </p>
  </div>
</div>
`;
